### PR TITLE
Support remove/update operations in go dataplane

### PR DESCRIPTION
**Pull request by praveingk**

### CODE REVIEW

File: pkg/dataplane/server/dataplane.go [hunk 1]
```diff
@@ -80,6 +95,17 @@ func (d *Dataplane) AddListener(ln *listener.Listener) {
 	}()
 }
 
+// RemoveListener removes a listener.
+func (d *Dataplane) RemoveListener(name string) {
```
Comment (orozery) [inline pkg/dataplane/server/dataplane.go:99]: This is a bit confusing:
You have:
- `CreateListener` - creates a listening socket
- `AddListener` - bookkeeping + call `CreateListener` in a new go-routine
- `DeleteListener` - async close of a listening socket
- `RemoveListener` - bookkeeping + call do `DeleteListener`

Maybe you can merge each matching two?

File: pkg/dataplane/server/dataplane.go [hunk 2]
```diff
@@ -61,16 +61,31 @@ func (d *Dataplane) GetClusterHost(name string) (string, error) {
...
 
-// AddCluster adds a cluster to the map.
+// AddCluster adds/updates a cluster to the map.
 func (d *Dataplane) AddCluster(c *cluster.Cluster) {
 	d.clusters[c.Name] = c
 }
 
+// RemoveCluster adds a cluster to the map.
+func (d *Dataplane) RemoveCluster(name string) {
+	delete(d.clusters, name)
```
Comment (orozery) [inline pkg/dataplane/server/dataplane.go:71]: I just noticed that you need to add a lock to guard d.clusters (in all places where it is accessed).

**Review by orozery — CHANGES_REQUESTED** [review]
Can you search the code for "https://github.com/clusterlink-net/clusterlink/issues/218" and handle the 2 TODOs that will enable e2e testing?

File: pkg/dataplane/server/dataplane.go [hunk 3]
```diff
@@ -40,6 +41,8 @@ type Dataplane struct {
 	clusters           map[string]*cluster.Cluster
 	listeners          map[string]*listener.Listener
 	listenerEnd        map[string]chan bool
+	clustersLock       sync.Mutex
+	listenerLock       sync.Mutex
```
Comment (orozery) [inline pkg/dataplane/server/dataplane.go:45]: I don't think this is needed as I don't see any concurrent access to the listeners map.

File: pkg/dataplane/server/dataplane.go [hunk 4]
```diff
@@ -40,6 +41,8 @@ type Dataplane struct {
 	clusters           map[string]*cluster.Cluster
 	listeners          map[string]*listener.Listener
 	listenerEnd        map[string]chan bool
+	clustersLock       sync.Mutex
```
Comment (orozery) [inline pkg/dataplane/server/dataplane.go:44]: I think you're missing taking a lock on `GetClusterTarget` and `GetClusterHost`.

File: pkg/dataplane/server/dataplane.go [hunk 5]
```diff
@@ -61,25 +64,59 @@ func (d *Dataplane) GetClusterHost(name string) (string, error) {
...
+}
+
+// RemoveCluster adds a cluster to the map.
+func (d *Dataplane) RemoveCluster(name string) {
+	d.clustersLock.Lock()
+	delete(d.clusters, name)
+	d.clustersLock.Unlock()
+}
+
+// GetClusters returns the listeners map.
```
Comment (orozery) [inline pkg/dataplane/server/dataplane.go:81]:
```suggestion
// GetClusters returns the clusters map.
```

**Review by orozery — CHANGES_REQUESTED** [review]

File: pkg/dataplane/server/dataplane.go (hunk 4 above)
Comment (praveingk) [reply pkg/dataplane/server/dataplane.go]: Done

File: pkg/dataplane/server/dataplane.go (hunk 3 above)
Comment (praveingk) [reply pkg/dataplane/server/dataplane.go]: Listener is also added and removed similar to clusters.

File: pkg/dataplane/server/dataplane.go (hunk 2 above)
Comment (elevran) [reply pkg/dataplane/server/dataplane.go:82]: you could potentially encapsulate the map+RWLock (for clusters and listeners, separately) and then add the Update/Delete methods on the new type with internal locking. For Getting the map, you may want to create a slice copy of the names and return that to the caller or handle the clearing internally via a method

File: pkg/dataplane/server/dataplane.go [hunk 6]
```diff
@@ -54,24 +59,50 @@ func (d *Dataplane) GetClusterTarget(name string) (string, error) {
...
+		if ln.Address.GetSocketAddress().GetAddress() == le.Address.GetSocketAddress().GetAddress() &&
+			ln.Address.GetSocketAddress().GetPortValue() == le.Address.GetSocketAddress().GetPortValue() {
+			d.listenerLock.RUnlock()
+			return
+		}
+		d.listenerEnd[listenerName] <- true
 	}
+	d.listenerLock.RUnlock()
+	d.listenerLock.Lock()
+	defer d.listenerLock.Unlock()
```
Comment (orozery) [inline pkg/dataplane/server/dataplane.go:105]: This seems inefficient to me.
I would take a writer lock at `handleClusters`, and `handleListeners`, and remove all locks from `AddListener`, 'AddCluster`, 'RemoveListener`, `RemoveCluster`.
Just add a comment for each function that says the function assumes lock is already held.

File: tests/k8s.sh [hunk 7]
```diff
@@ -80,7 +80,7 @@ function test_k8s {
   kubectl cp $SCRIPT_DIR/../pkg/policyengine/policytypes/examples/allowAll.json gwctl:/tmp/allowAll.json
   kubectl exec -i gwctl -- gwctl create policy --type access --policyFile /tmp/allowAll.json
 
-  # get imported service port
+  # get imported service port
```
Comment (orozery) [inline tests/k8s.sh:83]: nit: revert?

**Review by orozery — APPROVED** [review]

**Pull request merged**